File: src/chapters/numerical-calculus/content/code/richardson.cpp
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;

// Central-difference first derivative D(h), error O(h^2).
double central(function<double(double)> f, double x, double h) {
    return (f(x + h) - f(x - h)) / (2 * h);
}

int main() {
    cout.precision(10);
    cout << fixed;
    auto f = [](double x) { return sin(x); };
    double x = 1, h = 0.1;
    double d1 = central(f, x, h);        // D(h),   error O(h^2)
    double d2 = central(f, x, h / 2);    // D(h/2), error O(h^2)
    double ext = (4 * d2 - d1) / 3;      // Richardson extrapolation, error O(h^4)
    cout << "D(h)         = " << d1 << "\n";
    cout << "D(h/2)       = " << d2 << "\n";
    cout << "extrapolated = " << ext << "  exact cos(1) = " << cos(1.0) << "\n";
}

File: src/chapters/numerical-calculus/content/code/richardson.rs
// Central-difference first derivative D(h), error O(h^2).
fn central<F: Fn(f64) -> f64>(f: &F, x: f64, h: f64) -> f64 {
    (f(x + h) - f(x - h)) / (2.0 * h)
}

fn main() {
    let f = |x: f64| x.sin();
    let (x, h) = (1.0, 0.1);
    let d1 = central(&f, x, h);          // D(h),   error O(h^2)
    let d2 = central(&f, x, h / 2.0);    // D(h/2), error O(h^2)
    let ext = (4.0 * d2 - d1) / 3.0;     // Richardson extrapolation, error O(h^4)
    println!("D(h)         = {:.10}", d1);
    println!("D(h/2)       = {:.10}", d2);
    println!("extrapolated = {:.10}  exact cos(1) = {:.10}", ext, 1.0_f64.cos());
}
